server/src/actors/player/message.rs

```diff
@@ -1,49 +1,83 @@
 use actix::{Addr, AsyncContext, Context, Handler, Message};
 use chrono::{DateTime, Utc};
-use uuid::Uuid;
+use std::time::Instant;
 
 use super::super::api;
 use super::super::api::ApiClient;
 use super::super::server;
-use super::Player;
+use super::{Player, PlayerStatus};
 
 #[derive(Message)]
 #[rtype(result = "()")]
 pub struct Connected {
     pub api_client: Addr<ApiClient>,
-    pub session_id: Uuid,
 }
 
 impl Handler<Connected> for Player {
     type Result = ();
 
     fn handle(&mut self, msg: Connected, ctx: &mut Context<Self>) -> Self::Result {
-        if let None = self.api_client {
-            self.api_client = Some(msg.api_client.clone());
-            msg.api_client
-                .do_send(api::outbound_message::UserConnected {
-                    session_id: msg.session_id,
-                    username: self.username.clone(),
-                    player: ctx.address(),
-                });
-        } else {
-            msg.api_client
-                .do_send(api::outbound_message::ConnectionFailed {
-                    reason: "User still connected.".to_string(),
-                });
+        match self.status {
+            PlayerStatus::Connecting => {
+                self.status = PlayerStatus::Connected {
+                    api_client: msg.api_client.clone(),
+                };
+                msg.api_client
+                    .do_send(api::outbound_message::UserConnected {
+                        session_id: self.session_id.clone(),
+                        username: self.username.clone(),
+                        player: ctx.address(),
+                    });
+                self.server
+                    .do_send(server::message::GlobalChatBroadcast::system_message(
+                        format!("User '{}' connected.", self.username),
+                    ));
+            }
+            PlayerStatus::LostConnection { .. } => {
+                self.status = PlayerStatus::Connected {
+                    api_client: msg.api_client.clone(),
+                };
+                msg.api_client
+                    .do_send(api::outbound_message::UserConnected {
+                        session_id: self.session_id.clone(),
```

Comment:
I still feel like you could do without these `String` clones. But it might be tricky and thus not worth the effort :man_shrugging:

Reply:
`session_id` is `Uuid`, but I wanted all actors to own their data to avoid problems with lifetime in async context.

Reply:
true...not sure but I believe that such lifetimes could be a mess - I mean, these data are pretty small so it won't be a big deal (it's just that all these `.clone()`s kinda make it less elegant - you know, more text to read :) )

Reply:
Well you would need to ensure, that there is "global store" that owns the data and you have reference to it. My idea was to have all actor *independent*, just have address to another actor.